File: ios/trezor_flutter/Sources/trezor_flutter/TrezorFlutterPlugin.swift
import Flutter
import UIKit

/// iOS side of `trezor_flutter`: a Bluetooth LE packet pipe.
///
/// iOS has no USB path to a Trezor (no MFi accessory support), so the USB
/// methods report `unsupported`. The channel contract is the one documented in
/// `trezor_platform.dart` and implemented identically on Android.
public class TrezorFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var sink: FlutterEventSink?
  private lazy var ble = TrezorBleCentral { [weak self] event in
    // CoreBluetooth runs on the main queue (see TrezorBleCentral), so this is
    // already the thread FlutterEventSink requires.
    self?.sink?(event)
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = TrezorFlutterPlugin()
    let methods = FlutterMethodChannel(
      name: "trezor_flutter/methods", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: methods)
    let events = FlutterEventChannel(
      name: "trezor_flutter/events", binaryMessenger: registrar.messenger())
    events.setStreamHandler(instance)
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    sink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    sink = nil
    return nil
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    let deviceId = args?["deviceId"] as? String

    switch call.method {
    case "getCapabilities":
      // Reporting BLE support must not create the central manager: that is
      // what triggers the Bluetooth permission prompt.
      result(["usb": false, "ble": true])
    case "bluetoothState":
      result(ble.stateString())
    case "usbListDevices":
      result([])
    case "usbRequestPermission":
      result(FlutterError(code: "unsupported", message: "USB is not available on iOS", details: nil))
    case "bleStartScan":
      ble.startScan(result: result)
    case "bleStopScan":
      ble.stopScan()
      result(nil)
    case "open":
      guard let deviceId else { return missingId(result) }
      guard (args?["transport"] as? String) == "ble" else {
        return result(
          FlutterError(code: "unsupported", message: "Only Bluetooth is available on iOS", details: nil))
      }
      ble.open(deviceId: deviceId, result: result)
    case "write":
      guard let deviceId else { return missingId(result) }
      guard let data = args?["data"] as? FlutterStandardTypedData else {
        return result(FlutterError(code: "writeFailed", message: "Missing data", details: nil))
      }
      ble.write(deviceId: deviceId, data: data.data, result: result)
    case "close":
      guard let deviceId else { return missingId(result) }
      ble.close(deviceId: deviceId)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func missingId(_ result: FlutterResult) {
    result(FlutterError(code: "deviceNotFound", message: "Missing deviceId", details: nil))
  }
}

File: android/src/main/kotlin/io/github/mtpublic/trezor_flutter/TrezorFlutterPlugin.kt
package io.github.mtpublic.trezor_flutter

import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Android side of `trezor_flutter`: a packet pipe over USB or BLE.
 *
 * Nothing here knows the Trezor protocol. [TrezorUsbTransport] and
 * [TrezorBleTransport] find devices and move fixed-size packets; framing,
 * encryption and protobuf all happen in Dart. See `trezor_platform.dart` for
 * the channel contract, which the iOS plugin implements identically.
 *
 * Every event is posted to the main thread before it reaches the EventSink,
 * which Flutter requires; both transports call [emit] from their own threads.
 */
class TrezorFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    private companion object {
        const val METHOD_CHANNEL = "trezor_flutter/methods"
        const val EVENT_CHANNEL = "trezor_flutter/events"
    }

    private lateinit var context: Context
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var sink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private var usb: TrezorUsbTransport? = null
    private var ble: TrezorBleTransport? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL).also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL).also {
            it.setStreamHandler(this)
        }
        usb = TrezorUsbTransport(context, ::emit)
        ble = TrezorBleTransport(context, ::emit)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        methodChannel = null
        eventChannel = null
        usb?.dispose()
        ble?.dispose()
        usb = null
        ble = null
        sink = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    private fun emit(event: Map<String, Any?>) {
        mainHandler.post { sink?.success(event) }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val usb = this.usb
        val ble = this.ble
        if (usb == null || ble == null) {
            result.error("notAttached", "Plugin is detached", null)
            return
        }
        when (call.method) {
            "getCapabilities" -> {
                val pm = context.packageManager
                val bluetoothManager =
                    context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
                result.success(
                    mapOf(
                        "usb" to pm.hasSystemFeature(PackageManager.FEATURE_USB_HOST),
                        "ble" to (pm.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE) &&
                            bluetoothManager?.adapter != null),
                    )
                )
            }
            "bluetoothState" -> result.success(ble.adapterState())
            "usbListDevices" -> result.success(usb.listDevices())
            "usbRequestPermission" -> usb.requestPermission(requireId(call, result) ?: return, result)
            "bleStartScan" -> ble.startScan(result)
            "bleStopScan" -> {
                ble.stopScan()
                result.success(null)
            }
            "open" -> {
                val id = requireId(call, result) ?: return
                when (call.argument<String>("transport")) {
                    "usb" -> usb.open(id, result)
                    "ble" -> ble.open(id, result)
                    else -> result.error("openFailed", "Unknown transport", null)
                }
            }
            "write" -> {
                val id = requireId(call, result) ?: return
                val data = call.argument<ByteArray>("data")
                if (data == null) {
                    result.error("writeFailed", "Missing data", null)
                    return
                }
                when {
                    usb.isOpen(id) -> usb.write(id, data, result)
                    ble.isOpen(id) -> ble.write(id, data, result)
                    else -> result.error("notConnected", "Device $id is not open", null)
                }
            }
            "close" -> {
                val id = requireId(call, result) ?: return
                usb.close(id)
                ble.close(id)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun requireId(call: MethodCall, result: MethodChannel.Result): String? {
        val id = call.argument<String>("deviceId")
        if (id == null) result.error("deviceNotFound", "Missing deviceId", null)
        return id
    }
}
